### PR TITLE
logsafe for kvs-slow-log

### PR DESCRIPTION
**Pull request by samrogerson**
**Goals (and why)**: Wrap log events from `kvs-slow-log` in `logsafe` methods.

**Implementation Description (bullets)**: N/A

**Concerns (what feedback would you like?)**: Safety (or not) of `tableRef` and `method`

**Where should we start reviewing?**: `ProfilingKeyValueService`

**Priority (whenever / two weeks / yesterday)**: one week

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2222)

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java [hunk 1]
```diff
@@ -89,32 +91,38 @@ public static ProfilingKeyValueService create(KeyValueService delegate, long slo
             int numCells, long sizeInBytes) {
         return (logger, stopwatch) ->
                 logger.log("Call to KVS.{} on table {} for {} cells of overall size {} bytes took {} ms.",
-                        method, tableRef, numCells, sizeInBytes, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+                        SafeArg.of("method", method), UnsafeArg.of("tableRef", tableRef),
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: We can use `LoggingArgs.tableRef` here, which uses safe/unsafe depending on whether you've specified the table as safe

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java [hunk 2]
```diff
@@ -89,32 +91,38 @@ public static ProfilingKeyValueService create(KeyValueService delegate, long slo
...
     }
 
     private static BiConsumer<LoggingFunction, Stopwatch> logTimeAndTableRange(String method,
             TableReference tableRef,
             RangeRequest range) {
         return (logger, stopwatch) ->
                 logger.log("Call to KVS.{} on table {} with range {} took {} ms.",
-                        method, tableRef, range, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+                        SafeArg.of("method", method), UnsafeArg.of("tableRef", tableRef), SafeArg.of("range", range),
+                        stopwatch.elapsed(TimeUnit.MILLISECONDS));
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: should the time also be a `SafeArg` for all these?

**Comment by samrogerson** [comment]
Sorry it took me so long to get back to this - didn't see the comment notification.

Thoughts

- For the [`batchColumnRangeSelection`](https://github.com/palantir/atlasdb/pull/2222/files#diff-19049cb73b8b93d748412488ec39d3daR468) and [`columnRangeSelection`](https://github.com/palantir/atlasdb/pull/2222/files#diff-19049cb73b8b93d748412488ec39d3daR487) should I add a util to `LoggingArgs` to see if it's safe to log?
- Doing a lot of `(Un)SafeArg.of("STRING",...)` could probably put these somewhere like `LoggingArgs` to avoid magic strings but was following other usage style of safe-logging in e.g. `AbstractTransactionManager`

**Comment by hsaraogi** [comment]
Hey Sam, we have work in progress to have a way for tablenames to be be marked safe/unsafe. Using LoggingArgs will make sure that if the table is marked as safe it gets mapped to SafeArgs orElse unsafeArgs. Note that applications will have to modify their schema or tableMetadata to mark the table names as safe (unless they contain sensitive data) before they actually start showing up in logs.

**Comment by samrogerson** [comment]
Yup - have moved the table refs over. Was also wondering about
columnselections and the like :)

**Comment by nziebart** [comment]
Hey @samrogerson, yeah I think having a util for checking loggability of the range selections make sense - thanks!

**Comment by samrogerson** [comment]
Done on both counts

**Comment by samrogerson** [comment]
Tests added.  Should be good for another review @nziebart @hsaraogi

File: atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java [hunk 3]
```diff
@@ -68,7 +76,69 @@ static synchronized void setLogArbitrator(KeyValueServiceLogArbitrator arbitrato
                 logArbitrator.isColumnNameSafe(tableReference, columnName));
     }
 
+    public static Arg<Long> duration(Stopwatch stopwatch) {
+        return getArg("duration", stopwatch.elapsed(TimeUnit.MILLISECONDS), true);
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java]: nit: name this `durationMillis`

File: atlasdb-client/src/test/java/com/palantir/atlasdb/logging/LoggingArgsTest.java [hunk 4]
```diff
@@ -133,4 +163,59 @@ public void canReturnUnsafeColumnNameEvenIfTableReferenceIsSafe() {
...
+    public void returnsSafeRangeWhenAllSafe() {
+        assertThat(LoggingArgs.range(SAFE_TABLE_REFERENCE, SAFE_RANGE_REQUEST))
+                .isInstanceOf(SafeArg.class);
+    }
+
+    @Test
+    public void returnsUnsafeRangeWhenAllColumnsUnsafe() {
+        assertThat(LoggingArgs.range(SAFE_TABLE_REFERENCE, UNSAFE_RANGE_REQUEST))
+                .isInstanceOf(UnsafeArg.class);
+
```
Comment (nziebart) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/logging/LoggingArgsTest.java]: nit: empty line

File: atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java [hunk 5]
```diff
@@ -68,7 +76,69 @@ static synchronized void setLogArbitrator(KeyValueServiceLogArbitrator arbitrato
...
+
+    public static Arg<Integer> rowCount(int rowCount) {
+        return getArg("tableCount", rowCount, true);
+    }
+
+    public static Arg<Long> sizeInBytes(long sizeInBytes) {
+        return getArg("sizeInBytes", sizeInBytes, true);
+    }
+
+    public static Arg<Integer> columnCount(ColumnSelection columnSelection) {
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java]: could this method do the logic for specifying `"all"` vs a count? seems off to have a function called `allColumns` below that actually returns a count

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java [hunk 6]
```diff
@@ -356,10 +360,10 @@ public void dropTables(Set<TableReference> tableRefs) {
                         logger.log(
                                 "Call to KVS.getRows on table {} requesting {} columns from {} rows took {} ms ",
                                 LoggingArgs.tableRef(tableRef),
-                                columnSelection.allColumnsSelected() ? "all"
-                                        : Iterables.size(columnSelection.getSelectedColumns()),
-                                Iterables.size(rows),
-                                stopwatch.elapsed(TimeUnit.MILLISECONDS)),
+                                columnSelection.allColumnsSelected() ? LoggingArgs.allColumns()
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java]: see other comment about pushing this logic into `LoggingArgs.columnCount`

File: atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java (hunk 5 above)
Comment (samrogerson) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java]: all columns doesn't return a count, it returns `Arg<String>`.
Problem for `columnCount` is handling the return type `Arg<String>` vs `Arg<Integer>`.  Can just `toString` the integer but that felt weird

File: atlasdb-client/src/test/java/com/palantir/atlasdb/logging/LoggingArgsTest.java (hunk 4 above)
Comment (samrogerson) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/logging/LoggingArgsTest.java]: done

File: atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java (hunk 3 above)
Comment (samrogerson) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java]: done

File: atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java (hunk 5 above)
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java]: oh i see.. how about just returning `Arg<?>`?

Comment (samrogerson) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/logging/LoggingArgs.java]: done

**Pull request merged**